rfcs/0128-redeployable-clients.md

```diff
@@ -0,0 +1,393 @@
+support.
+
+## Changes to HTTP references format
+
+The `api-reference.json` document currently served
+http://schemas.taskcluster.net/base/v1/api-reference.json from should be served
+from `<TASKCLUSTER_ROOT_URL>/schemas/base/v1/http-reference.json`.
+
+* All `entries[*].input` and `entries[*].output` properties should be URLs
+  relative to `<taskclusterRootURL>/schemas/<serviceName>`
+
+(This has already been implemented across most services)
+
+## Changes to AMQP references format
+
+The `api-reference.json` document currently served
+http://schemas.taskcluster.net/base/v1/api-reference.json from should be served
+from `<TASKCLUSTER_ROOT_URL>/schemas/base/v1/amqp-reference.json`.
+
+* All `entries[*].schema` properties should be URLs relative to
+  `<taskclusterRootURL>/schemas/<serviceName>`
+
+(This has already been implemented across most services)
+
+## Changes to publication of API references and schemas
+
+The implementation must serve the described resources under the given URLs set
+out in this document. The author is not concerned with how a service declares
+its API references to the platform during build/deploy stages in order that the
+references are included in the API references manifest, that is a matter for
+the implementor.
+
+The author considers it reasonable though, that there could be an API endpoint
+that enables services to register their API references with the taskcluster
+platform when they start up. An advantage of this approach (rather than burning
+this information into the build/deploy steps) is that a taskcluster platform
+can evolve, with parties able to declare new services at runtime, that aren't
+necessarily able to affect the taskcluster build/deploy steps.
+
+For example, in an organisation where there is a taskcluster deployment team
+looking after a deployment, they may wish to grant scopes to another team who
+are developing additional services which plug into the core taskcluster
+deployment. By granting this other team the scopes to declare API references,
+the team can work autonomously in developing new services, without the
+taskcluster deployment team needing to restart or rebuild their taskcluster
+deployment. The second team can make their API references available and modify
+them as they see fit while the taskcluster core platform is running without any
+maintenance overhead.
+
+But we may wish to implement such a thing in a future PR - for the time being,
+the author does not care about the publish mechanism, only that the content is
+eventually served.
+
+## Changes to taskcluster client building procedure
+
+### All language clients (go, java, node.js, python)
+
+The build process for these clients should be moved into a standalone tool that
+requires a `TASKCLUSTER_ROOT_URL` to be specified to build the client for. It
+will then query the API References manifest in order to determine all the APIs
+it needs to build, and from there build packages for each API reference it
+finds.
```

Comment:
I'm a bit confused about what's being proposed here, between this section and the next few.  My understanding of the latest plan, which seems to agree with *most* of the text here, is this:

 * All client packages uploaded to hosting services like npm or maven are mostly http clients, and contain only a default set of API references built from the latest upstream Taskcluster build, but no generated code
 * Client consumers are expected to provide API references or use the default bundled set of API references
    * for some languages, there is a code-generation process
    * for some languages, the API references are simply provided at runtime
    * (maybe some languages support both options)
 * We recommend vendoring (that is, including in the repo with the client consumer) either the API references or the resulting generated code, or both.  This means that changes to the APIs against which the software is built occur in a version-control-visible way.  But of course users can choose to "re-vendor" on every build if they like.

With that arrangement, the tool mentioned here is basically responsible only for downloading the manifests and references and storing them in a data format that can subsequently be checked into a repo.  Does it also preform code generation for the various languages?  Is it going to be difficult to set those languages up so that they find the right architecture binary to do their code generation?  If so, it might make sense for the code generation to be written in the target languages, instead.